_posts/2021-04-22-torch-for-optimization/torch-for-optimization.Rmd

```diff
@@ -0,0 +1,640 @@
+
+### L-BFGS (!)
+
+Among the many `torch` optimizers commonly used in deep learning (Adam, AdamW, RMSprop ...), there is one "outsider", much better known in classic numerical optimization than in neural-networks space: L-BFGS, a.k.a. [Limited-memory BFGS](https://en.wikipedia.org/wiki/Limited-memory_BFGS), a memory-optimized implementation of the [Broyden--Fletcher--Goldfarb--Shanno optimization algorithm](https://en.wikipedia.org/wiki/Broyden%E2%80%93Fletcher%E2%80%93Goldfarb%E2%80%93Shanno_algorithm) (BFGS).
+
+BFGS is perhaps the most widely used among the so-called Quasi-Newton, second-order optimization algorithms. As opposed to the family of first-order algorithms that, in deciding on a descent direction, make use of gradient information only, second-order algorithms additionally take curvature information into account. To that end, exact Newton methods actually compute the Hessian (a costly operation), while Quasi-Newton methods avoid that cost and, instead, resort to iterative approximation.
+
+Looking at the contours of the Rosenbrock function, with its prolonged, narrow valley, it is not difficult to imagine that curvature information might make a difference. And, as you'll see in a second, it really does. Before though, one note on the code. When using L-BFGS, it is necessary to wrap function call and gradient evaluation in a closure (`calc_loss()`, in the below snippet), for them to be callable several times per iteration. You can convince yourself that the closure is, in fact, entered repeatedly, by inspecting this code snippet's chatty output:
+
+```{r}
+num_iterations <- 3
+
+x_star <- torch_tensor(c(-1, 1), requires_grad = TRUE)
+
+optimizer <- optim_lbfgs(x_star)
+
+calc_loss <- function() {
+
+  optimizer$zero_grad()
+
+  value <- rosenbrock(x_star)
+  cat("Value is: ", as.numeric(value), "\n")
+
+  value$backward()
+  cat("Gradient is: ", as.matrix(x_star$grad), "\n\n")
+  value
+
+}
+
+for (i in 1:num_iterations) {
+  cat("Iteration: ", i, "\n")
+  optimizer$step(calc_loss)
+}
+
+```
+
+    Iteration:  1 
+    Value is:  4 
+    Gradient is:  -4 0 
+
+    Value is:  6 
+    Gradient is:  -2 10 
+
+    ...
+    ...
+
+    Value is:  0.04880721 
+    Gradient is:  -0.262119 -0.1132655 
+
+    Value is:  0.0302862 
+    Gradient is:  1.293824 -0.7403332 
+
+    Iteration:  2 
+    Value is:  0.01697086 
+    Gradient is:  0.3468466 -0.3173429 
+
+    Value is:  0.01124081 
+    Gradient is:  0.2420997 -0.2347881 
+
+    ...
+    ...
+
+    Value is:  1.111701e-09 
+    Gradient is:  0.0002865837 -0.0001251698 
+
+    Value is:  4.547474e-12 
+    Gradient is:  -1.907349e-05 9.536743e-06 
+
+    Iteration:  3 
+    Value is:  4.547474e-12 
+    Gradient is:  -1.907349e-05 9.536743e-06 
+
+Even though we ran the algorithm for three iterations, the optimal value really is reached after two. Seeing how well this worked, we try L-BFGS on a slightly more crazy function, named *flower*, for pretty self-evident reasons.
```

Comment:
```suggestion
Even though we ran the algorithm for three iterations, the optimal value really is reached after two. Seeing how well this worked, we try L-BFGS on a slightly more complex function, named *flower*, for pretty self-evident reasons.
```

Reply:
You could use a bunch of different words here (e.g. complex, complicated, elaborate…), but I'd avoid "crazy" both because it's unclear what that means, and because it reads as "ableist" these days: https://www.npr.org/2019/07/08/739643765/why-people-are-arguing-to-stop-using-the-words-crazy-and-insane

Reply:
yes, I get it... i guess "more difficult" (without the "slightly") would do as well...